### PR TITLE
Polygon Cleanup

### PR DESCRIPTION
**Pull request by foxox**
I have been using SolveSpace a lot to design things for my 3d printer recently and I'm a professional software engineer so I thought I would take a look at the code.
I cleaned up various things in the Polygon component... const correctness, project code format adherence, int types, pass by const reference for nonprimitives, readability improvements, etc.
These changes should make the cost more robust to bugs in the future and may even have some performance impact.
If you like this I may do the same with other files later.
I can squash commits after review if you like.

### CODE REVIEW

**Comment by whitequark** [comment]
Thank you for spending time to improve SolveSpace. However, I would ask you that you consult the maintainers before making far-reaching, sweeping changes.

There are a few serious issues with this PR that would prevent it from being merged, and which are general for any changes that can be applied:

  * Every commit should do one specific change with one specific purpose.
  * Every commit should have a commit message that clearly states why the change is done.
  * SolveSpace is a mature codebase that is written in a somewhat unorthodox dialect of C++. We do not change it for the sake of change or for the sake of making it slightly more elegant, since the amount of churn and bugs introduced due to typos and other simple mistakes outweighs the benefits.
  * In particular, as an extension of the above, SolveSpace does not have a large testsuite. We heavily rely on `git annotate` for determining what commit introduced some code, and use that for determining how it is supposed to work. Sweeping changes like modifying indentation, adding braces, etc break this completely, which would make further maintenance much harder.

Specific to this PR:
  * There is no good justification for making local variables (themselves) constant. In a better language (like Rust) this would be the default, but we already have our codebase in C++, and changing almost every single variable declaration does not justify the minimal (if non-zero at all) amount of bugs it would highlight.
  * There is absolutely no point in replacing `i++` with `++i` when `i` is POD. If there is a compiler where the latter is more efficient, then this compiler is explicitly not supported for building SolveSpace. Trivial optimizations can always be relied on happening. (I'm pretty sure this is also true for non-POD cases at this point, but I'd have to check.)

**Comment by foxox** [comment]
@whitequark Pull requests exist to facilitate this kind of conversation -- thank you. In my day job, a 300 line cleanup PR to one component is nothing... gets reviewed by 2 or 3 other people, they leave a few comments, I push commits in response, get approval, squash the commits into one, it gets merged, and everyone is glad for it. I don't often contribute to open source projects, so your response was a bit surprising. Since you maintain this project, I'll concede on most points since they don't matter.

However, I disagree about const correctness. Const correctness should be applied to all code, even local variables. If a variable is not meant to be mutated after it is initialized, it should be const. It makes the code easier to read and easier to change because it shows the author's intent and (as you mentioned) helps catch bugs. Rust is irrelevant... `const` is a core C++ feature and it's there to be used. C++ coding standards like HIC++ or the JSF standard or whatever typically include a requirement for const correctness in all situations.

Also, "churn" isn't a problem if code follows best practices and is documented properly. Would things be different if there were more unit tests in this project?

**Comment by whitequark** [comment]
> In my day job, a 300 line cleanup PR to one component is nothing... gets reviewed by 2 or 3 other people, they leave a few comments, I push commits in response, get approval, squash the commits into one, it gets merged, and everyone is glad for it. I don't often contribute to open source projects, so your response was a bit surprising.

This is common in certain open source projects with a large team of developers typically paid by top tech companies maintaining them, for example, LLVM. However, SolveSpace has one maintainer--me--I am not paid (in fact I'm beyond broke at this point), and I have health issues that limit the amount of time I can spend on FOSS. I definitely do not have the bandwidth to review thousands of lines of cleanups that just check a box in some coding standard.

> C++ coding standards like HIC++ or the JSF standard or whatever typically include a requirement for const correctness in all situations.

Coding standards are not a gospel but a tool in a large array of tools that improve code quality. SolveSpace does not and will not blindly adhere to any particular coding standard or rule unless there's demonstrable evidence that following it (and changing code to follow it) does more good than harm.

If you look through the git log, there is a number of commits that justify a few sweeping changes in great detail, e.g. the commit that changed the initialization style. I expect any cleanups to justify themselves at least that well, including any churn they cause.

> Also, "churn" isn't a problem if code follows best practices and is documented properly. Would things be different if there were more unit tests in this project?

I've spent something like 1.5 years learning how to maintain this project without breaking it, and about half of my time was spent fixing bugs that were the direct result of various "cleanups" and "consistency improvements". This codebase is sparsely tested--almost untested, really--and has a very large amount of hidden invariants. Your approach is unfortunately not applicable to a legacy codebase like this.

In short, my direct experience doing what you are trying to do in this PR results in my understanding that it is undesirable and infeasible to do while removing more bugs than adding them. This, of course, could change if this codebase would become well-documented and well-tested to the point where any trivial errors introduced in bulk refactoring would be caught by the test suite. That is not the case today.

**Comment by whitequark** [comment]
> However, I disagree about const correctness. Const correctness should be applied to all code, even local variables. If a variable is not meant to be mutated after it is initialized, it should be const. It makes the code easier to read and easier to change because it shows the author's intent and (as you mentioned) helps catch bugs.

To expand on this, I agree with you in principle, and, although I do not use this style in my other C++ projects, I would be open to using it in a greenfield codebase. But I don't think that const correctness for locals justifies a commit that would break `git annotate` for almost the entire codebase; `git annotate` is a tool tool that is critical to understanding the code, so it's not going to happen.

**Comment by Evil-Spirit** [comment]
@foxox const correctness is something you can apply for projects created from scratch. Even in this case, you can't link any third-party libraries without pain. Your programming skills and time will start working on const-correctness maintaining only. If sometime someone show you what really means to follow some code style _exactly_ how it written, you will be scared about typing every space and newline. You will be disarmed completely. My way was a hard - I was stucked inside premature optimizations, const-correctness, template magic, and so on. Pure programming without result. But I worked with some people who crossed throgh this. Now I don't bother about optimization - there is profiler to show me what's wrong with, I don't bother because of codestyle - I can follow any codestyle. I don't bother about programming language - I can use any. I can blame any code and make it more optimal in some context - make it faster, make it shorter, make it more flexible, or more understandable. But I can't do it simultaneously.

**Comment by whitequark** [comment]
@Evil-Spirit Thanks, I completely forgot about the issue with 3rd party libraries, which we've already hit.

**Comment by whitequark** [comment]
@bcmpinc Thank you, this review reflects my thoughts as well.

File: src/polygon.cpp [hunk 1]
```diff
@@ -472,10 +517,12 @@ int SKdNodeEdges::AnyEdgeCrossings(Vector a, Vector b, int cnt,
...
         for(sell = edges; sell; sell = sell->next) {
-            SEdge *se = sell->se;
-            if(se->tag == cnt) continue;
+            SEdge * const se = sell->se;
+            if(se->tag == cnt) {
+                continue;
+            }
             if(se->EdgeCrosses(a, b, pi, spl)) {
-                inters++;
+                ++inters;
```
[the remark this answers is not in the document]
Comment (foxox) [reply src/polygon.cpp]: It's not a matter of outsmarting the compiler. Pre-increment is the preferred operation in this context because it is simpler, which makes it more readable/understandable and less prone to errors introduced in the future. Ex: imagine if something like this would need to be done in the future: `const size_t num_inters_this_iteration = inters++;` It's easy to make this kind of bug when post-increment is used habitually. It's a shame that the language wasn't called ++C

File: src/polygon.cpp [hunk 2]
```diff
@@ -522,13 +569,13 @@ void SPointList::Clear() {
...
+bool SPointList::ContainsPoint(const Vector& pt) const {
     return (IndexForPoint(pt) >= 0);
 }
 
-int SPointList::IndexForPoint(Vector pt) const {
-    int i;
-    for(i = 0; i < l.n; i++) {
+int SPointList::IndexForPoint(const Vector& pt) const {
+    size_t i;
+    for(i = 0; i < l.n; ++i) {
```
[the remark this answers is not in the document]
Comment (foxox) [reply src/polygon.cpp]: Hmm, this function should probably initialize `int i` with `-1` and return it at the end

File: src/polygon.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (whitequark) [reply src/polygon.cpp]: I don't agree that pre-increment is simpler or more readable in this context (indeed, since no other codebase I work on uses pre-increment pervasively, it trips me up each time), but more importantly, you are not considering the unintended consequences of making such a change globally. Namely:
  * it is not semantics-preserving, and so it will introduce bugs, and we don't have nearly enough test coverage to catch them;
  * it screws up `git blame`, and we don't have nearly enough documentation to not rely on that.

Therefore this change and changes similar to it are not useful, but are harmful.

File: src/polygon.h [hunk 3]
```diff
@@ -119,15 +122,15 @@ class SContour {
...
-    double SignedAreaProjdToNormal(Vector n) const;
-    bool IsClockwiseProjdToNormal(Vector n) const;
-    bool ContainsPointProjdToNormal(Vector n, Vector p) const;
-    void OffsetInto(SContour *dest, double r) const;
-    void CopyInto(SContour *dest) const;
+    double SignedAreaProjdToNormal(const Vector& n) const;
+    bool IsClockwiseProjdToNormal(const Vector& n) const;
+    bool ContainsPointProjdToNormal(const Vector& n, const Vector& p) const;
+    void OffsetInto(SContour *dest, const double r) const;
+    void CopyInto(SContour * const dest) const;
```
[the remark this answers is not in the document]
Comment (foxox) [reply src/polygon.h]: Isn't this basic C++?

[the remark this answers is not in the document]
Comment (whitequark) [reply src/polygon.h]: No, it is nothing but visual noise in the header (the constness of the argument inside the function has no bearing whatsoever on its public interface; yet another design mistake in C++).

File: src/polygon.h [hunk 4]
```diff
@@ -180,14 +184,14 @@ class STriangle {
         Vector normals[3];
     };
 
-    static STriangle From(STriMeta meta, Vector a, Vector b, Vector c);
+    static STriangle From(const STriMeta& meta, const Vector& a, const Vector& b, const Vector& c);
```
[the remark this answers is not in the document]
Comment (foxox) [reply src/polygon.h]: What's the cutoff? How big does a type need to be before it's worthwhile to pass by reference/pointer instead of by value? I can find a number of references online indicating that it's good practice in C++ to pass any non-primitive type by reference. Has this stuff been profiled?

[the remark this answers is not in the document]
Comment (whitequark) [reply src/polygon.h]: > Has this stuff been profiled?

Yes, and we've converted most functions on the hot paths to pass POD classes that are larger than a pointer to pass-by-const-reference.

File: src/polygon.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Evil-Spirit) [reply src/polygon.h]: i think it's like writing `void foo(const int param)`. No one writing code like this.

**Comment by whitequark** [comment]
I believe this kind of cleanup is unproductive and so I am going to close this PR. The reasons are outlined above multiple times and I don't have anything to add to that.

I would, of course, be happy to collaborate with you on other changes that take into account the requirements imposed by the project's history and maintenance needs.
